File: api/game.py
import random

from master import models as master
from api import events
from api.exceptions import *
from api.models import Question, Choice, AnsweredQuestion, GameTurn

MIN_PLAYERS_TO_START = 2


def init(game):
    mquestions = list(master.Question.objects.filter(lang=game.lang))
    mchoices = list(master.Choice.objects.filter(lang=game.lang))

    if not mquestions:
        raise NoQuestionFound

    if not mchoices:
        raise NoChoiceFound

    random.shuffle(mquestions)
    random.shuffle(mchoices)

    questions = list(map(lambda mq: Question(game=game, text=mq.text), mquestions))
    Question.objects.bulk_create(questions)

    for idx, question in enumerate(questions):
        for place in mquestions[idx].get_blanks():
            question.blanks.create(place=place)

    choices = map(lambda mc: Choice(game=game, text=mc.text), mchoices)
    Choice.objects.bulk_create(choices)

    events.game_created(game.owner)


def start(game):
    if game.players.count() < MIN_PLAYERS_TO_START:
        raise NotEnoughPlayers

    game.state = 'started'
    game.question_master = random.choice(game.players.all())

    game.deal_cards()
    game.pick_question()
    game.save()

    events.game_started(game)


def next_turn(game, question_master):
    game.question_master = question_master

    game.deal_cards()
    game.pick_question()
    game.save()

    events.next_turn(game)


def answer(game, choices, answered_by):
    question = game.current_question
    answered_question = AnsweredQuestion(game=game, question=question, answered_by=answered_by)
    answered_question.save()

    blanks = question.blanks.order_by('place')

    for i in range(len(choices)):
        choice = choices[i]
        blank = None

        if question.blanks.count() > 0:
            blank = blanks[i]

        answered_question.answers.create(position=blank, choice=choice)

        choice.owner = None
        choice.played = True
        choice.save()

    events.answer_submitted(game, answered_by)

    if game.get_propositions().count() == game.players.count() - 1:
        all_answers_submitted(game)

    return answered_question


def all_answers_submitted(game):
    answers = list(game.get_propositions())
    random.shuffle(answers)

    for idx, answer in enumerate(answers):
        answer.place = idx
        answer.save()

    events.all_answers_submitted(game, answers)


def select_answer(game, selected, selected_by):
    winner = selected.answered_by

    selected.selected_by = selected_by
    selected.save()

    turns_count = game.turns.count()
    turn = GameTurn(
        number=turns_count + 1,
        game=game,
        question_master=game.question_master,
        winner=winner,
        question=game.current_question,
    )
    turn.save()

    for answer in game.get_propositions():
        answer.turn = turn
        answer.save()

    events.answer_selected(game, turn)

File: api/src/game.js
const { Sequelize, Choice, Answer } = require('./models');
const events = require('./events');
const Op = Sequelize.Op;

function shuffle(a) {
  let j, x, i;

  for (i = a.length - 1; i > 0; i--) {
    j = Math.floor(Math.random() * (i + 1));
    x = a[i];
    a[i] = a[j];
    a[j] = x;
  }

  return a;
}

async function join(game, player) {
  await game.addPlayer(player);
  events.emit('game:join', game, player);
}

async function leave(game, player) {
  await player.removeCards(await player.getCards());
  await game.removePlayer(player);
  events.emit('game:leave', game, player);
}

async function dealCards(game, player) {
  const choices = await game.getChoices({
    where: { available: true },
    limit: game.cardsPerPlayer - await player.countCards(),
  });

  await Choice.update({ playerId: player.id, available: false }, {
    where: {
      id: { [Op.in]: choices.map(c => c.id) },
    },
  });

  return choices;
}

async function pickQuestion(game) {
  const question = (await game.getQuestions({
    where: { available: true },
    limit: 1,
  }))[0];

  if (!question)
    return false;

  await question.update({ available: false });
  await game.setCurrentQuestion(question);

  return true;
}

async function start(game) {
  const qm = (await game.getPlayers({
    order: Sequelize.fn('RANDOM'),
    limit: 1,
  }))[0];

  await game.createCards();
  await pickQuestion(game);
  await game.setQuestionMaster(qm);

  const players = await game.getPlayers();

  for (let i = 0; i < players.length; ++i) {
    const player = players[i];
    const choices = await dealCards(game, player);

    events.emit('player:cards', player, choices, { initial: true });
  }

  await game.update({ state: 'started' });

  events.emit('game:start', game);
}

async function answer(game, player, choices) {
  const answer = await game.createAnswer({
    playerId: player.id,
    questionId: game.questionId,
  });

  await Choice.update({ answerId: answer.id, playerId: null }, {
    where: {
      id: { [Op.in]: choices.map(c => c.id) },
    },
  });

  if (await game.getPlayState() === 'question_master_selection') {
    const propositions = shuffle(await game.getPropositions());

    for (let i = 0; i < propositions.length; ++i)
      await propositions[i].update({ place: i + 1 });
  }

  events.emit('game:answer', game, player, answer);
}

async function select(game, player, answer) {
  await game.setSelectedAnswer(answer);
  events.emit('game:select', game, player, answer);
}

async function nextTurn(game) {
  const players = await game.getPlayers();
  const answer = await game.getSelectedAnswer({ include: ['player'] });
  const winner = answer.player;

  const turn = await game.createTurn({
    number: await game.countTurns() + 1,
    questionId: game.questionId,
    questionMasterId: game.questionMasterId,
    winnerId: winner.id,
  });

  await Answer.update({
    gameTurnId: turn.id,
  }, {
    where: { questionId: game.questionId },
  });

  events.emit('game:turn', game, turn);

  await game.setSelectedAnswer(null);
  if (!await pickQuestion(game)) {
    await end(game);
    return;
  }

  await game.setQuestionMaster(winner);

  for (let i = 0; i < players.length; ++i) {
    const player = players[i];
    const choices = await dealCards(game, player);

    if (choices.length > 0)
      events.emit('player:cards', player, choices);
  }

  events.emit('game:next', game);
}

async function end(game) {
  await game.update({
    state: 'finished',
    questionMasterId: null,
    selectedAnswerId: null,
    questionId: null,
  });

  events.emit('game:end', game);
}

module.exports = {
  join,
  leave,
  start,
  dealCards,
  pickQuestion,
  answer,
  select,
  nextTurn,
  end,
};
